### PR TITLE
Service Account patching

### LINKED ISSUE #3300: [Bug] Custom Service Account being overwritten by operator
**Describe the bug**
Before 0.18.0 if the KafkaConnect service account already existed it would just use it, now it is overwriting it with the strimzi-managed metadata.

**To Reproduce**
1. Create a ServiceAccount `some-name-connect` with any custom annotation and/or labels
2. Create a KafkaConnect `some-name`, any configuration works
3. The ServiceAccount metadata will be replaced

**Expected behavior**
The ServiceAccount should be left unchanged like before.

**Environment (please complete the following information):**
 - Strimzi version: 0.18.0
 - Installation method: Helm Chart
 - Kubernetes cluster: v1.14.10-gke.36
 - Infrastructure: GKE

**Additional context**
We use custom annotations on the service account to allow identity sharing from the cluster to the external resources (external to the cluster) using [GKE's Workload Identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity)

### LINKED ISSUE #3707: [Enhancement] Support for Service Account annotations
**Is your feature request related to a problem? Please describe.**
I need a way to annotate Service Accounts created by Strimzi's Kafka Operator.

**Describe the solution you'd like**
Ideally, I would like to see support for the `serviceaccount` construct under the `template` attribute of the various Spec Template schemas such as KafkaConnectSpecTemplate, KafkaMirrorMaker2SpecTemplate etc. in the same way `deployment`, `pod`, etc. are currently supported.

**Describe alternatives you've considered**
#3179 contains what I would consider a "hack" solution: create the Service Account myself out of band/outside Strimzi first with the exact name Strimzi will use such that this SA is used instead of Strimzi creating its own. This seem brittle to break Strimzi's encapsulation and subject to breaking if Strimzi's K8s resource naming scheme were to ever change.

**Additional context**
Just like the author of #3179, we need this specifically for use by AWS EKS's [IAM for Service Accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-technical-overview.html) feature although this would be useful for anyone with any need to annotate a Strimzi managed Service Account.

Thank you

### PR DESCRIPTION
**Pull request by scholzj**
### Type of change

- Enhancement / new feature

### Description

This PR adds the service account patching as described in the [proposal 026](https://github.com/strimzi/proposals/blob/main/026-service-account-patching.md). It adds new feature gate `ServiceAccountPatching` and when it is enabled, it patches the SAs when needed in every reconciliation. It also adds possibility to configure service account metadata in the `template` sections.

This should close #3707 and close #3300

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md

### CODE REVIEW

File: api/src/main/java/io/strimzi/api/kafka/model/template/KafkaBridgeTemplate.java [hunk 1]
```diff
@@ -87,6 +88,16 @@ public void setBridgeContainer(ContainerTemplate bridgeContainer) {
         this.bridgeContainer = bridgeContainer;
     }
 
+    @Description("Template for the Kafka Bridge Service Account.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/template/KafkaBridgeTemplate.java:91]:
```suggestion
    @Description("Template for the Kafka Bridge service account.")
```

File: api/src/main/java/io/strimzi/api/kafka/model/template/KafkaClusterTemplate.java [hunk 2]
```diff
@@ -226,6 +227,16 @@ public void setClusterRoleBinding(ResourceTemplate clusterRoleBinding) {
         this.clusterRoleBinding = clusterRoleBinding;
     }
 
+    @Description("Template for the Kafka Service Account.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/template/KafkaClusterTemplate.java:230]:
```suggestion
    @Description("Template for the Kafka service account.")
```

File: api/src/main/java/io/strimzi/api/kafka/model/template/KafkaConnectTemplate.java [hunk 3]
```diff
@@ -145,6 +147,26 @@ public void setClusterRoleBinding(ResourceTemplate clusterRoleBinding) {
         this.clusterRoleBinding = clusterRoleBinding;
     }
 
+    @Description("Template for the Kafka Connect Service Account.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/template/KafkaConnectTemplate.java:150]:
```suggestion
    @Description("Template for the Kafka Connect service account.")
```

File: api/src/main/java/io/strimzi/api/kafka/model/template/KafkaConnectTemplate.java [hunk 4]
```diff
@@ -145,6 +147,26 @@ public void setClusterRoleBinding(ResourceTemplate clusterRoleBinding) {
...
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
+    public ResourceTemplate getServiceAccount() {
+        return serviceAccount;
+    }
+
+    public void setServiceAccount(ResourceTemplate serviceAccount) {
+        this.serviceAccount = serviceAccount;
+    }
+
+    @Description("Template for the Kafka Connect Build Service Account.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/template/KafkaConnectTemplate.java:160]:
```suggestion
    @Description("Template for the Kafka Connect Build service account.")
```

File: api/src/main/java/io/strimzi/api/kafka/model/template/KafkaExporterTemplate.java [hunk 5]
```diff
@@ -81,6 +82,16 @@ public void setContainer(ContainerTemplate container) {
         this.container = container;
     }
 
+    @Description("Template for the Kafka Exporter Service Account.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/template/KafkaExporterTemplate.java:85]:
```suggestion
    @Description("Template for the Kafka Exporter service account.")
```

File: api/src/main/java/io/strimzi/api/kafka/model/template/KafkaMirrorMaker2Template.java [hunk 6]
```diff
@@ -87,6 +88,16 @@ public void setMirrorMaker2Container(ContainerTemplate mirrorMaker2Container) {
         this.mirrorMaker2Container = mirrorMaker2Container;
     }
 
+    @Description("Template for the Kafka Mirror Maker 2 Service Account.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/template/KafkaMirrorMaker2Template.java:91]:
```suggestion
    @Description("Template for the Kafka MirrorMaker 2 service account.")
```

File: api/src/main/java/io/strimzi/api/kafka/model/template/KafkaMirrorMakerTemplate.java [hunk 7]
```diff
@@ -76,6 +76,16 @@ public void setMirrorMakerContainer(ContainerTemplate mirrorMakerContainer) {
         this.mirrorMakerContainer = mirrorMakerContainer;
     }
 
+    @Description("Template for the Kafka Mirror Maker Service Account.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/template/KafkaMirrorMakerTemplate.java:79]:
```suggestion
    @Description("Template for the Kafka MirrorMaker service account.")
```

File: api/src/main/java/io/strimzi/api/kafka/model/template/ZookeeperClusterTemplate.java [hunk 8]
```diff
@@ -127,6 +128,16 @@ public void setTlsSidecarContainer(ContainerTemplate tlsSidecarContainer) {
         this.tlsSidecarContainer = tlsSidecarContainer;
     }
 
+    @Description("Template for the ZooKeeper Service Account.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/template/ZookeeperClusterTemplate.java:131]:
```suggestion
    @Description("Template for the ZooKeeper service account.")
```

File: api/src/main/java/io/strimzi/api/kafka/model/template/CruiseControlTemplate.java [hunk 9]
```diff
@@ -99,6 +100,16 @@ public void setTlsSidecarContainer(ContainerTemplate tlsSidecarContainer) {
         this.tlsSidecarContainer = tlsSidecarContainer;
     }
 
+    @Description("Template for the Cruise Control Service Account.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/template/CruiseControlTemplate.java:103]:
```suggestion
    @Description("Template for the Cruise Control service account.")
```

File: api/src/main/java/io/strimzi/api/kafka/model/template/EntityOperatorTemplate.java [hunk 10]
```diff
@@ -87,6 +87,16 @@ public void setTlsSidecarContainer(ContainerTemplate tlsSidecarContainer) {
         this.tlsSidecarContainer = tlsSidecarContainer;
     }
 
+    @Description("Template for the Entity Operator Service Account.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/template/EntityOperatorTemplate.java:90]:
```suggestion
    @Description("Template for the Entity Operator service account.")
```

File: api/src/main/java/io/strimzi/api/kafka/model/template/JmxTransTemplate.java [hunk 11]
```diff
@@ -65,6 +66,16 @@ public void setContainer(ContainerTemplate container) {
         this.container = container;
     }
 
+    @Description("Template for the JMX Trans Service Account.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/template/JmxTransTemplate.java:69]:
```suggestion
    @Description("Template for the JMX Trans service account.")
```

**Review by PaulRMellor — APPROVED** [review]
Looks good.

**Comment by PaulRMellor** [comment]
@scholzj  I've suggested lower case _service account_ in the Java descriptions. Going with Kubernetes naming convention. Otherwise `ServiceAccount`?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/FeatureGatesTest.java [hunk 12]
```diff
@@ -18,8 +18,13 @@
     @ParallelTest
     public void testFeatureGates() {
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/FeatureGatesTest.java:19]: I think we should add a method to `FeatureGates` (if it doesn't already have it) to get the known gates, and an assertion here for the known gates. That way it's impossible to add a gate and not update this test.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperatorMockTest.java [hunk 13]
```diff
@@ -119,7 +120,7 @@ public void afterEach() {
                 new DefaultAdminClientProvider(),
                 new DefaultZookeeperScalerProvider(),
                 ResourceUtils.metricsProvider(),
-                pfa, 60_000L);
+                pfa, new FeatureGates(""), 60_000L);
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperatorMockTest.java:123]: Maybe add a `FeatureGates.NONE` constant?

**Review by tombentley — APPROVED** [review]
A couple of nits, but LGTM.

**Pull request merged**